crates/accelerate/src/gate_direction.rs

```diff
@@ -0,0 +1,159 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2024
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use crate::nlayout::PhysicalQubit;
+use crate::target_transpiler::{Qargs, Target};
+use pyo3::prelude::*;
+use pyo3::types::{PySet, PyTuple};
+use qiskit_circuit::imports;
+use qiskit_circuit::operations::{OperationRef, PyInstruction};
+use qiskit_circuit::{
+    dag_circuit::{DAGCircuit, NodeType},
+    error::DAGCircuitError,
+    operations::Operation,
+    packed_instruction::PackedInstruction,
+    Qubit,
+};
+
+/// Check if the two-qubit gates follow the right direction with respect to the coupling map.
+///
+/// Args:
+///     dag: the DAGCircuit to analyze
+///
+///     coupling_edges: set of edge pairs representing a directed coupling map, against which gate directionality is checked
+///
+/// Returns:
+///     true iff all two-qubit gates comply with the coupling constraints
+#[pyfunction]
+#[pyo3(name = "check_gate_direction_coupling")]
+fn py_check_with_coupling_map(
+    py: Python,
+    dag: &DAGCircuit,
+    coupling_edges: &Bound<PySet>,
+) -> PyResult<bool> {
+    let coupling_map_check =
+        |curr_dag: &DAGCircuit, _: &PackedInstruction, op_args: &[Qubit]| -> bool {
+            coupling_edges
+                .contains((
+                    map_qubit(py, dag, curr_dag, op_args[0]).0,
+                    map_qubit(py, dag, curr_dag, op_args[1]).0,
+                ))
+                .unwrap_or(false)
+        };
+
+    check_gate_direction(py, dag, &coupling_map_check)
+}
+
+/// Check if the two-qubit gates follow the right direction with respect to instructions supported in the given target.
+///
+/// Args:
+///     dag: the DAGCircuit to analyze
+///
+///     target: the Target against which gate directionality compliance is checked
+///
+/// Returns:
+///     true iff all two-qubit gates comply with the target's coupling constraints
+#[pyfunction]
+#[pyo3(name = "check_gate_direction_target")]
+fn py_check_with_target(py: Python, dag: &DAGCircuit, target: &Bound<Target>) -> PyResult<bool> {
+    let target = target.borrow();
+
+    let target_check =
+        |curr_dag: &DAGCircuit, inst: &PackedInstruction, op_args: &[Qubit]| -> bool {
+            let mut qargs = Qargs::new();
+
+            qargs.push(PhysicalQubit::new(
+                map_qubit(py, dag, curr_dag, op_args[0]).0,
+            ));
+            qargs.push(PhysicalQubit::new(
+                map_qubit(py, dag, curr_dag, op_args[1]).0,
+            ));
+
+            target.instruction_supported(inst.op.name(), Some(&qargs))
```

Comment:
I'm not a fan that `instruction_supported` requires a `SmallVec` reference, I feel like it could work with a slice input. But we can save that for a follow up. But given these constraints I think you can do or something like this to avoid the new/push since `Qargs` is just a type alias:

```suggestion
            let qargs = smallvec![
                PhysicalQubit::new(
                    map_qubit(py, dag, curr_dag, op_args[0]).0,
                ),
                PhysicalQubit::new(
                    map_qubit(py, dag, curr_dag, op_args[1]).0,
                ),
            ];
            target.instruction_supported(inst.op.name(), Some(&qargs))
```

Reply:
Done.